### PR TITLE
[incubator/logstash] Revamp logstash from Deployment to StatefulSet to support Persisted Queues

### PR DESCRIPTION
**Pull request by ghost**
* Give boilerplate for kafka input and output.
* Prepare for persisted queues use case by default.
* Configure logstash via env var instead of configmap.
* Prepare for real HTTP ingress.
* Disable legacy syslog protocols by default, keep boilerplate.
* Nominate `beats` as the only default input.
* Support exposing non-HTTP TCP ports via separate LoadBalancer service.
* Parameterize many prod-centric settings.
* Support easy Prometheus scraping of logstash_exporter running in each pod.
  * https://github.com/BonnierNews/logstash_exporter
* Upgrade to logstash-oss:6.2.2
* Use logstash health API for logstash health checks
* Optionally seeds data at Prometheus for this awesome Logstash Grafana dashboard
  * https://grafana.com/dashboards/2525
* Set default pipeline to beats->ES
* Include official boilerplate for ES output (when beats input)
* Describe multi-pipeline usage recommendations.
* Show NOTES.txt for HTTP only when applicable.
* Reverse erroneous addition of josdotso as maintainer.
  * (Would like to revisit in the future, once I become more experienced with the repo and chart.)
  * If you are actively seeking maintainers, please let me know. I work with some other folks who may wish to work toward that goal with me.
* Add notes on best practice of loading beats-generated.
  index-template into elasticsearch.
* Refactored AWS-specific annotations as examples at values file.
  * Give example of service annotation to realize AWS Route53 DNS record for internal (or otherwise) ELB.
  * Give example of how to make serviceLoadBalancer an internal AWS ELB.
* Implemented storageClass best practices at values and statefulset.

### CODE REVIEW

**Comment by ghost** [comment]
screen shot 2018-02-23 at 10 28 45 am

^ chaoskube vs. Kubernetes rescheduling a logstash pod. view is from aforementioned Grafana dashboard.

**Comment by unguiculus** [comment]
cc @christian-roggia

File: incubator/logstash/Chart.yaml [hunk 1]
```diff
@@ -15,3 +15,5 @@ maintainers:
   email: jrodgers@powerhrg.com
 - name: christian-roggia
   email: christian.roggia@gmail.com
+- name: josdotso
```
Comment (unguiculus) [inline incubator/logstash/Chart.yaml]: Please clarify why you added yourself as maintainer. You will be expected to continue contributing and review PRs.

@christian-roggia Are you fine with this?

File: incubator/logstash/templates/service-aws-lb-internal.yaml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{{- if .Values.serviceAwsLoadBalancerInternal.enabled }}
```
Comment (unguiculus) [inline incubator/logstash/templates/service-aws-lb-internal.yaml]: The service should not be AWS-specific.

File: incubator/logstash/templates/statefulset.yaml [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        - name: pipeline
+          configMap:
+            name: {{ template "logstash.fullname" . }}-pipeline
+
+  volumeClaimTemplates:
+    - metadata:
+        name: data
+      spec:
+        accessModes: ["ReadWriteOnce"]
+        storageClassName: {{ .Values.persistentVolumes.storageClass }}
```
Comment (unguiculus) [inline incubator/logstash/templates/statefulset.yaml]: Please follow our best practices for storage class configuration.

See #1869.

File: incubator/logstash/templates/statefulset.yaml [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      {{- if .Values.podLabels }}
+        # Custom pod labels
+        {{- range $key, $value := .Values.podLabels }}
+        {{ $key }}: {{ $value | quote }}
+        {{- end }}
+      {{- end }}
+      annotations:
+        checksum/patterns: {{ include (print $.Template.BasePath "/patterns-config.yaml") . | sha256sum }}
+        checksum/pipeline: {{ include (print $.Template.BasePath "/pipeline-config.yaml") . | sha256sum }}
+      {{- if .Values.exporter.logstash.enabled }}
```
Comment (unguiculus) [inline incubator/logstash/templates/statefulset.yaml]: Please make these annotations configurable separately. You don't need them if you use the prometheus-operator.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thank you for this excellent contribution. I especially appreciate updating the chart to currrent best practices.

File: incubator/logstash/Chart.yaml (hunk 1 above)
Comment (ghost) [reply incubator/logstash/Chart.yaml]: Thanks @unguiculus!  This was my rookie mistake. I hope to continue contributions to this and other charts in the repo, but I'd should become more comfortable with the patterns and processes before committing to being a binding voice in reviews.  Once I have some more PRs behind me, perhaps we can revisit?

cc: @christian-roggia

File: incubator/logstash/templates/service-aws-lb-internal.yaml (hunk 2 above)
Comment (ghost) [reply incubator/logstash/templates/service-aws-lb-internal.yaml]: Fixed in latest push.

File: incubator/logstash/templates/statefulset.yaml (hunk 3 above)
Comment (ghost) [reply incubator/logstash/templates/statefulset.yaml]: Should be fixed in latest push.  Thanks!

File: incubator/logstash/templates/statefulset.yaml (hunk 4 above)
Comment (ghost) [reply incubator/logstash/templates/statefulset.yaml]: These are now just commented-out example annotations at values.yaml.  Please let me know if this is agreeable.  Thanks!

**Comment by ghost** [comment]
Thanks again, @unguiculus!  I updated the top comment and responded to comments inline after pushing new commits.

Here's a view of Grafana Logstash Dashboard showing Beats events coming through and going out to Elasticsearch.
screen shot 2018-02-23 at 11 15 47 pm

**Comment by ghost** [comment]
FYI: I had to amend my last commit as my local build of the chart slipped past me and into the commit.  Sorry about that!

File: incubator/logstash/templates/service-lb.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{{- if .Values.serviceLoadBalancer.enabled }}
```
Comment (unguiculus) [inline incubator/logstash/templates/service-lb.yaml]: Why do you need a second service? This seems to duplicate the existing one that can already be configured to be of type LoadBalancer if desired, which I think is better anyways.

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: incubator/logstash/templates/service-lb.yaml (hunk 5 above)
Comment (ghost) [reply incubator/logstash/templates/service-lb.yaml]: Yes. This is a good point and I agree. I've deleted the second service and configured the first service to accept annotations passed from values.yaml.   Thanks!

**Review by unguiculus — APPROVED** [review]

**Comment by unguiculus** [comment]
Let's give @christian-roggia some more time to review.

**Comment by christian-roggia** [comment]
I apologize but unfortunately I won't be able to review PRs for around a week, I am moving to a new house and my workplace is not ready yet.

I suggest to proceed with the PR anyway if it looks good to @unguiculus, I will test and review everything as soon as I will be able to.

**Comment by rendhalver** [comment]
Thanks!
This will give me a nice intro to StatefulSets.

**Pull request merged**